hail/src/main/scala/is/hail/asm4s/FunctionBuilder.scala

```diff
@@ -16,6 +15,185 @@ import scala.collection.generic.Growable
+
+  def genField[T: TypeInfo](): Field[T] = newField[T](genName("f"))
+
+  def genField[T: TypeInfo](suffix: String): Field[T] = newField(genName("f", suffix))
+
+  def classAsBytes(print: Option[PrintWriter] = None): Array[Byte] = {
+    init.instructions.add(new InsnNode(RETURN))
+
+    val cw = new ClassWriter(ClassWriter.COMPUTE_MAXS + ClassWriter.COMPUTE_FRAMES)
+    val sw1 = new StringWriter()
+    var bytes: Array[Byte] = new Array[Byte](0)
+    try {
+      for (method <- cn.methods.asInstanceOf[util.List[MethodNode]].asScala) {
+        val count = method.instructions.size
+        log.info(s"instruction count: $count: ${ cn.name }.${ method.name }")
+        if (count > 8000)
+          log.warn(s"big method: $count: ${ cn.name }.${ method.name }")
+      }
+
+      cn.accept(cw)
+      bytes = cw.toByteArray
+      //       This next line should always be commented out!
+      //      CheckClassAdapter.verify(new ClassReader(bytes), false, new PrintWriter(sw1))
+    } catch {
+      case e: Exception =>
+        // if we fail with frames, try without frames for better error message
+        val cwNoFrames = new ClassWriter(ClassWriter.COMPUTE_MAXS)
+        val sw2 = new StringWriter()
+        cn.accept(cwNoFrames)
+        try {
+          CheckClassAdapter.verify(new ClassReader(cwNoFrames.toByteArray), false, new PrintWriter(sw2))
+        } catch {
+          case e: Exception =>
+            log.error("Verify Output 1 for " + name + ":")
+            throw e
+        }
+
+        if (sw2.toString.length() != 0) {
+          System.err.println("Verify Output 2 for " + name + ":")
+          System.err.println(sw2)
+          throw new IllegalStateException("Bytecode failed verification 1", e)
+        } else {
+          if (sw1.toString.length() != 0) {
+            System.err.println("Verify Output 1 for " + name + ":")
+            System.err.println(sw1)
+          }
+          throw e
+        }
+    }
+
+    if (sw1.toString.length != 0) {
+      System.err.println("Verify Output 1 for " + name + ":")
+      System.err.println(sw1)
+      throw new IllegalStateException("Bytecode failed verification 2")
+    }
+
+    print.foreach { pw =>
+      val cr = new ClassReader(bytes)
+      val tcv = new TraceClassVisitor(null, new Textifier, pw)
+      cr.accept(tcv, 0)
+    }
+    bytes
+  }
+
+  def result(print: Option[PrintWriter] = None): () => C = {
+    val childClasses = children.result().map(f => (f.name.replace("/","."), f.classAsBytes(print)))
+
+    val bytes = classAsBytes(print)
+    val n = name.replace("/",".")
+
+    assert(TaskContext.get() == null,
+      "FunctionBuilder emission should happen on master, but happened on worker")
+
+    new (() => C) with java.io.Serializable {
+      @transient
+      @volatile private var theClass: Class[_] = null
+
+      def apply(): C = {
+        try {
+          if (theClass == null) {
+            this.synchronized {
```

Comment:
Not a complaint, but why is the `synchronized` here necessary? Are we doing anything multi threaded?

Reply:
Yes, the executors are multi-threaded.

Reply:
But FYI the driver is not.